### PR TITLE
Rewrite &= -> = and simplify the initial blocking phase.

### PR DESCRIPTION
**Pull request by oon3m0oo**
@fenrus75 it looked to me like the initial read spin could be combined in the loop with the write, since it was actually checking more values than it needed to, but correct me if my logic is wrong.

### CODE REVIEW

File: driver/level3/level3_thread.c [hunk 1]
```diff
@@ -462,14 +485,16 @@ static int inner_thread(blas_arg_t *args, BLASLONG *range_m, BLASLONG *range_n,
...
 #ifdef TIMING
           ops += 2 * min_i * MIN(range_n[current + 1]  - js, div_n) * min_l;
 #endif
-          
+
           /* Clear synchronization flag if this thread is done with region of B */
           if (is + min_i >= m_to) {
+            ANNOTATE_IGNORE_WRITES_BEGIN();
             job[current].working[mypos][CACHE_LINE_SIZE * bufferside] &= 0;
+            ANNOTATE_IGNORE_WRITES_END();
```
Comment (fenrus75) [inline driver/level3/level3_thread.c]: the &= 0 in the line above is odd btw, it makes the write LESS atomic if notihing else.
should just really be an assignment

**Review by fenrus75 — COMMENTED** [review]
If you use LLVM and know it enough... would you mind putting the appropriate compiler barriers for MB/WMB in place? without those I can see the analysis getting double confused

**Comment by martin-frbg** [comment]
Seems you would need to include <valgrind/drd.h> (or its contents) for this to work ? Not sure if that would be worth it, just to suppress a warning from one specific tool.

**Comment by oon3m0oo** [comment]
@martin-frbg it's up to you.  Worst case we can have a local patch on our side that adds the annotations.

@fenrus75 The &= 0 has been there for some time, I actually assumed you had put it in there knowing some secret Intel sauce!  If the right thing to do here is not have the annotations but still move the loop and eliminate the &= 0 I can do that here, instead.

**Comment by martin-frbg** [comment]
That &=0 line seems to come from the original GotoBLAS code...
(Undecided about the annotations, but in their current form they fail all CI tests

**Comment by fenrus75** [comment]
&= 0 is basically equal to = 0

but because of the atomics /volatile I think it'll make the compiler
generate worse, less atomic code.

so I did not touch it before, but I think it should be = 0 so that we don't
have read-modifiy-write just write.

**Comment by oon3m0oo** [comment]
Alright this is now the bare minimum.  I'll set up a local process for adding the annotations.

**Pull request merged**